### PR TITLE
51 setup config protocol

### PR DESCRIPTION
**Pull request by hk-ret**
Handling setup configuration when DB is not available including initialization of DB.
(currently unittests not available

### CODE REVIEW

File: backend/src/persistance/bo_descriptors.py [hunk 1]
```diff
@@ -71,10 +88,53 @@ def validate(self, value):
...
+
+
+class BODict(_PersistantAttr):
+    @classmethod
+    def data_type(cls):
+        return dict
+
+    def validate(self, value):
+        try:
+            json.dumps(value, separators=(",", ":"))
```
Comment (DrDiHai) [inline backend/src/persistance/bo_descriptors.py:107]: Should also validate that the value parameter is of type dictionary

File: backend/src/persistance/bo_descriptors.py [hunk 2]
```diff
@@ -71,10 +88,53 @@ def validate(self, value):
...
+
+
+class BOList(_PersistantAttr):
+    @classmethod
+    def data_type(cls):
+        return list
+
+    def validate(self, value):
+        try:
+            json.dumps(value, separators=(",", ":"))
```
Comment (DrDiHai) [inline backend/src/persistance/bo_descriptors.py:120]: Should also validate that the value parameter is a list

**Review by DrDiHai — CHANGES_REQUESTED** [review]

File: backend/src/core/status.py [hunk 3]
```diff
@@ -10,11 +11,14 @@
 
 class Status(StrEnum):
     STATUS_UNCONFIGURED = "unconfigured"
+    STATUS_ONLINE = "online"
```
[the remark this answers is not in the document]
Comment (hk-ret) [reply backend/src/core/status.py]: 👍

File: backend/src/data/management/configuration.py [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+from core.app_logging import getLogger
+
+LOG = getLogger(__name__)
+
+
+class Configuration(BOBase):
+    user_id = BORelation(flag_values={"relation": User})
+    configuration = BODict()
+
+    def __init__(self, id: int = None, cfg: any = None, user_id: int = None) -> None:
```
[the remark this answers is not in the document]
Comment (hk-ret) [reply backend/src/data/management/configuration.py]: working on it

File: backend/src/data/management/configuration.py [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+        super().__init__(id=id)
+        self.user_id = user_id
+        self.configuration = cfg
+
+    def __repr__(self) -> str:
```
[the remark this answers is not in the document]
Comment (hk-ret) [reply backend/src/data/management/configuration.py:34]: working on this too

File: backend/src/data/management/user.py [hunk 6]
```diff
@@ -8,16 +10,26 @@
 LOG = getLogger(__name__)
 
 
+class UserRole(StrEnum):
```
[the remark this answers is not in the document]
Comment (hk-ret) [reply backend/src/data/management/user.py:13]: 👍

File: backend/src/database/db.py
[the remark this answers is not in the document]
Comment (hk-ret) [reply backend/src/database/db.py:1]: 👍

File: backend/src/messages/data.py [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+
+    @classmethod
+    def message_type(cls):
+        return MessageType.WS_TYPE_OBJECT
+
+    def __init__(
+        self,
+        object_type: DataObjectTypes,
+        index: int | str,
+        payload: any,
```
[the remark this answers is not in the document]
Comment (hk-ret) [reply backend/src/messages/data.py:42]: 👍

File: backend/src/messages/login.py [hunk 8]
```diff
@@ -25,21 +30,52 @@ def message_type(cls):
...
+        )
         token = self.message.get(MessageAttribute.WS_ATTR_TOKEN)
-        session = Session.get_session_from_token(
-            ses_token=self.message.get(MessageAttribute.WS_ATTR_SES_TOKEN),
-            conn_token=self.message.get(MessageAttribute.WS_ATTR_PREV_TOKEN),
-        ) or (Session(user, token, connection) if user else None)
-        if session:
+        if single_user_mode:
+            user = User(name=user_name)
+        else:
```
[the remark this answers is not in the document]
Comment (hk-ret) [reply backend/src/messages/login.py:45]: working on it

File: backend/src/messages/login.py [hunk 9]
```diff
@@ -25,21 +30,52 @@ def message_type(cls):
...
+                await connection.send_message(
+                    ByeMessage(reason="Password not matching")
+                )
+                LOG.debug("login failed (password)")
+                return
+        try:
+            session = (
+                Session.get_session_from_token(
+                    ses_token=self.message.get(MessageAttribute.WS_ATTR_SES_TOKEN),
+                    conn_token=self.message.get(MessageAttribute.WS_ATTR_PREV_TOKEN),
```
[the remark this answers is not in the document]
Comment (hk-ret) [reply backend/src/messages/login.py:57]: 👍

File: backend/src/money_pilot.py [hunk 10]
```diff
@@ -17,21 +19,46 @@ async def main():
...
+                        # LOG.debug("DB NOT available.")
+                        App.db_failure.set()
                     LOG.info(f"App running. (Status: {App.status})")
+                    await App.db_request_restart.wait()
+                    App.db_request_restart.clear()
 
-                    await asyncio.Future()
             except DBRestart:
-                LOG.info("DB restarted")
+                LOG.warning("DB Restart Exception")
```
[the remark this answers is not in the document]
Comment (hk-ret) [reply backend/src/money_pilot.py:49]: 👍

File: backend/src/persistance/bo_descriptors.py (hunk 1 above)
Comment (hk-ret) [reply backend/src/persistance/bo_descriptors.py:107]: 👍

File: backend/src/persistance/bo_descriptors.py (hunk 2 above)
Comment (hk-ret) [reply backend/src/persistance/bo_descriptors.py:120]: 👍

File: backend/src/messages/login.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (hk-ret) [reply backend/src/messages/login.py:45]: Now user is only checked if no session or prev token is provided in multi user mode

File: backend/src/data/management/configuration.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (hk-ret) [reply backend/src/data/management/configuration.py]: 👍

File: backend/src/data/management/configuration.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (hk-ret) [reply backend/src/data/management/configuration.py]: 👍

File: backend/src/database/sqlite.py [hunk 11]
```diff
@@ -93,8 +150,17 @@ def row_factory(cursor, row):
...
+        if not db_path.parent.exists():
+            LOG.info(f"Create missing directory '{db_path.parent}' for SQLite DB.")
+            db_path.parent.mkdir(parents=True)
+        if not db_path.parent.is_dir():
+            raise FileExistsError(
+                f"Path containing SQLite DB exists and is not a directory: {db_path.parent}"
+            )
         self._connection = await aiosqlite.connect(
-            database=self._cfg[Config.CONFIG_DB_FILE]
+            database=db_path, detect_types=sqlite3.PARSE_DECLTYPES
```
[the remark this answers is not in the document]
Comment (hk-ret) [reply backend/src/database/sqlite.py:163]: 👍

**Review by DrDiHai — APPROVED** [review]

**Pull request merged**